# Comparing posts from Hacker News
We will be comparing posts from hacker news, and seeing if posts that start with 'Ask HN' (posts that ask a question) have more interactivity(more comments) than posts with 'Show HN' (Posts that show a project or something interesting)

In [1]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[0:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [2]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts), len(show_posts), len(other_posts))
print(ask_posts[0:5])

1744 1162 17194
[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [3]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
    
total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
    
avg_ask_comments = total_ask_comments/len(ask_posts)
avg_show_comments = total_show_comments/len(show_posts)

print('average ask comments is {avg1:.2f} and the average show comments is {avg2:.2f}'.format(avg1 = avg_ask_comments,
                                                                                      avg2 = avg_show_comments))


average ask comments is 14.04 and the average show comments is 10.32


# Results
The average amount of comments is the ask posts is 14.04, and the average amount is the  show comments is 10.32. It seems like that ask posts garner more activity than show posts.

In [4]:
import datetime as dt
result_list = []

for row in ask_posts:
    result_list.append([row[6], int(row[4])])
counts_by_hours = {}
comments_by_hours = {}
for row in result_list:
    hour = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    hour = hour.strftime('%H')
    if hour in counts_by_hours:
        counts_by_hours[hour] += 1
        comments_by_hours[hour] += row[1]
    else:
        counts_by_hours[hour] = 1
        comments_by_hours[hour] = row[1]
print(counts_by_hours)
print(comments_by_hours)

{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [5]:
avg_by_hour = []

for key in counts_by_hours:
    average = [key]
    average.append(round(comments_by_hours[key]/counts_by_hours[key],2))
    avg_by_hour.append(average)
print(avg_by_hour)

[['09', 5.58], ['13', 14.74], ['10', 13.44], ['14', 13.23], ['16', 16.8], ['23', 7.99], ['12', 9.41], ['17', 11.46], ['15', 38.59], ['21', 16.01], ['20', 21.52], ['02', 23.81], ['18', 13.2], ['03', 7.8], ['05', 10.09], ['19', 10.8], ['01', 11.38], ['22', 6.75], ['08', 10.25], ['04', 7.17], ['00', 8.13], ['06', 9.02], ['07', 7.85], ['11', 11.05]]


In [6]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
sorted_swap= sorted(swap_avg_by_hour, reverse=True)
print('Top 5 hours for Ask Posts Comments!')
template = "{hour}:00: {avg:.2f} average comments per post."

print(hour)

for row in sorted_swap[0:5]:
    print(template.format(hour = row[1], avg= row[0]))
    

Top 5 hours for Ask Posts Comments!
09
15:00: 38.59 average comments per post.
02:00: 23.81 average comments per post.
20:00: 21.52 average comments per post.
16:00: 16.80 average comments per post.
21:00: 16.01 average comments per post.


# Results 
The list below shows the top 5 hours to post a 'Ask Post' on Hacker News
(All in eastern time)

1. 15:00: 38.59 average comments per post.

2. 02:00: 23.81 average comments per post.

3. 20:00: 21.52 average comments per post.

4. 16:00: 16.80 average comments per post.

5. 21:00: 16.01 average comments per post.

